internal/migrations/changesets.go

```diff
@@ -0,0 +1,319 @@
+	// "changeset_migration" is the event type used for replication of changesets from old chain to new chain during migration.
+	ChangesetMigrationEventType = "changeset_migration"
+)
+
+var (
+	migrationCfg *MigrationConfig
+
+	ErrNoMoreChunksToRead = errors.New("no more chunks to read")
+)
+
+func init() {
+	err := resolutions.RegisterResolution(ChangesetMigrationEventType, resolutions.ModAdd, ChangesetMigrationResolution)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// During network migration, the new node starts with the genesis state equal to the state of the old chain at the migration start height.
+// The changeset migration system is used to replicate changesets from the old chain to the new chain in order to
+// sync the state changes that occur on the old chain after the migration start height.
+// These changesets are applied in the order of the block heights starting from the migration start height.
+// Since the changesets can be very large and these changesets are to be sent through the voting system in the resolution body,
+// and are constrained by the block sizes, the changesets are split into chunks and sent through the voting system.
+// ChangesetMigration is the struct that represents the changeset migration chunk.
+type ChangesetMigration struct {
+	// Height is the block height the changeset belongs to.
+	Height *big.Int
+
+	// ChangesetIdx is the index of the changeset chunk in the block.
+	// Indexes starts from 0.
+	ChunkIdx *big.Int
+
+	// TotalChunks is the total number of chunks in the changeset.
+	TotalChunks *big.Int
+
+	// Changeset is the serialized changeset chunk.
+	Changeset []byte
+}
+
+// MarshalBinary marshals the ChangesetMigration into a binary format.
+func (cs *ChangesetMigration) MarshalBinary() ([]byte, error) {
+	return serialize.Encode(cs)
+}
+
+// UnmarshalBinary unmarshals the ChangesetMigration from a binary format.
+func (cs *ChangesetMigration) UnmarshalBinary(data []byte) error {
+	return serialize.Decode(data, cs)
+}
+
+// ChangesetMigrationResolution is the definition for changeset migration vote type in Kwil's voting system.
+// ChangesetMigrationResolution is responsible for applying changesets from the old chain to the new chain.
+// Once the changeset chunk is approved through the voting system, the chunk is added to the database.
+// Once all the chunks for a particular height are received, the changeset is applied and the chunks are deleted.
+// The changesets are applied in the order of block heights.
+var ChangesetMigrationResolution = resolutions.ResolutionConfig{
+	ConfirmationThreshold: big.NewRat(2, 3),
+	ExpirationPeriod:      24 * 60 * 10, // 1 day in blocks
+	ResolveFunc: func(ctx context.Context, app *common.App, resolution *resolutions.Resolution, block *common.BlockContext) error {
+
+		// Extract the migration config from the app, executed only once
+		if migrationCfg == nil {
+			migrationCfg = &MigrationConfig{}
+			cfg, ok := app.Service.ExtensionConfigs[ListenerName]
```

Comment:
Feels like a big footgun, since a simple misconfiguration could lead to a validator forking themselves.

IMO, I think it would make more sense to include the previous migration state as part of the network's genesis data (maybe. as part of the snapshot). This reduces the risk of screwing up, and decreases the manual coordination needed to upgrade a network.